webboot/webboot.go

```diff
@@ -26,32 +33,102 @@ var (
+//io.ReadCloser and err is returned from wget
+// if the arg is a name of an URL found in the osSystems, the Download Link specificed to the name of ISO is used for wget & urlToName to generate a name
+// and the Download Link then is returned instead of the passed-in URL
+func wgetArgs(arg string) (io.ReadCloser, string, string, error) {
+	var URL, filename string
+	//checks if arg is a key in osSystems map & a valid name to an ISO
+	if u, ok := osSystems[arg]; ok {
+		//if found in map, it finds the ISO's download link to use as URL for wget
+		URL = u.DownloadLink
+		filename = arg
+	} else {
+		//calls urlToName to change the filename to be the base of the link
+		filename = urlToName(arg)
+	}
+	read, err := wget(URL)
+	return read, URL, filename, err
+}
+
+func wget(URL string) (io.ReadCloser, error) {
+	//gets a response from the specified URL
+	resp, err := http.Get(URL)
+	if err != nil {
+		return nil, err
+	}
+	//checks if request to receive URL succeeded and the transmission level is successful on http level
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("non-200 HTTP status: %d", resp.StatusCode)
+	}
+	return resp.Body, nil
+}
+
+//func writer copies from an io.ReadCloser into filename.
+func writer(read io.ReadCloser, filename string) (string, error) {
+	w, err := os.Create(filename)
+	if err != nil {
+		return "", err
+	}
+
+	defer w.Close()
+
+	_, err = io.Copy(w, read)
+	return filename, err
+}
+
+//func urlToName takes URL link & generates a filename using the base of URL link
+// if the URL is empty or URL Path ends in /, urlToName returns index.html
+func urlToName(URL string) string {
+	//if output file is not initialized, argURl is parsed to generate a name for the output file
+	url, err := url.Parse(URL)
+
+	if err != nil {
+		log.Fatalf("Error parsing the URL, got %v, want nil", err)
+	}
+	filename := "index.html"
+	// takes the last part of the url Path as the name of output file. e.g. "archlinux-2019.08.01-x86_64.iso"
+	if url.Path != "" && !strings.HasSuffix(url.Path, "/") {
+		filename = path.Base(url.Path)
+	}
+	return filename
+}
+
+func usage() {
+	log.Printf("Usage: %s [ARGS] URL or name of ISO\n", os.Args[0])
+	flag.PrintDefaults()
+	os.Exit(1)
+}
+
 func main() {
 
 	flag.Parse()
-
+	//allows user to only pass in one argument e.g. "webboot [nameOFISO]" or "webboot [URL]"
+	if flag.NArg() != 1 {
+		usage()
+	}
 	dhclient.Request(*ifName, *timeout, *retry, *verbose, *ipv4, *ipv6)
 
-	if len(flag.Args()) < 1 {
-		log.Fatal("error:pass in an operating system")
+	arg := strings.ToLower(flag.Arg(0))
+	read, URL, filename, err := wgetArgs(arg)
+	if err != nil {
+		log.Fatalf("Error downloading %q: %v", URL, err)
 	}
 
-	osname := flag.Args()[0]
-
-	if _, ok := osSystems[osname]; !ok {
-		log.Printf("error operating system is not supported,The following systems are supported")
-		for os := range osSystems {
-			log.Printf("%v", os)
-		}
-		log.Fatalf("exit")
+	_, err = writer(read, filename)
```

Comment:
you don't use the filename, so why does writer return it?

Reply:
Yes, I see what you are saying. I will only return an error from the writer function() since I do not use the filename